File: src/trie.cpp
#include <iostream>
#include <vector>
#include "trie.hpp"

Trie::Trie()
{
    this->isWord = false;

    for (int i=0; i < CHAR_SIZE; i++)
    {
        this->trieNode[i] = nullptr;
    }
}

void Trie::addWord(std::string word)
{
    Trie* cur = this;

    for(int i=0; i < word.length(); i++)
    {
        if (cur->trieNode[word[i]] == nullptr){
            cur->trieNode[word[i]] = new Trie();
        }

        cur = cur->trieNode[word[i]];
    }

    cur->isWord = true;
}

bool Trie::searchTrie(std::string word)
{
    Trie* cur = this;

    for(int i=0 ; i < word.length(); i++)
    {
        if (cur->trieNode[word[i]] == nullptr){
            return false;
        }
        cur = cur->trieNode[word[i]];
    }

    return cur->isWord;   
}

std::vector<std::string> Trie::wordsByPrefix(std::string prefix){
    Trie* cur = this;
    std::vector<std::string> words;

    for(int i=0; i< prefix.length(); i++)
    {
        if (cur->trieNode[prefix[i]] == nullptr){
            return words;
        }

        cur = cur->trieNode[prefix[i]];
    } 

    words = allWordsFromNode(cur, prefix);
    if (cur->isWord){
        words.insert(words.begin(), prefix);
    }
   
    return words;
}

std::vector<std::string> Trie::allWordsFromNode(Trie*& node, std::string prefix)
{
    std::vector<std::string> words;
    Trie* cur = node;

    for(int i=0; i < CHAR_SIZE; i++) 
    {
        if (cur->trieNode[i] != nullptr) {
            std::string curPrefix = prefix + (char)i;

            if (cur->trieNode[i]->isWord){
                words.push_back(curPrefix);
            }

            std::vector<std::string> tmp = allWordsFromNode(cur->trieNode[i], curPrefix);
            for (int o=0; o < tmp.size(); o++)
            {
                words.push_back(tmp[o]);
            }           
        }
    }

    return words;
}

std::vector<std::string> Trie::allWordsInTrie()
{
    Trie* cur = this;
    std::vector<std::string> words = allWordsFromNode(cur, "");

    return words;
}

File: prototype/trie.py
class TrieNode:
    def __init__(self, letter):
        self.letter = letter
        self.children = {}
        self.isWord = False
        self.full_word = None


class Trie:
    def __init__(self):
        self.root = TrieNode("*")

    def add_word(self, word):
        cur = self.root
        for letter in word:
            if letter not in cur.children:
                cur.children[letter] = TrieNode(letter)

            cur = cur.children[letter]

        cur.isWord = True
        cur.full_word = word

    def check_word(self, word):
        cur = self.root

        for letter in word:
            if letter not in cur.children:
                return False
            cur = cur.children[letter]

        return cur.isWord

    def all_words_from_node(self, start_node):
        words = []

        for key, val in start_node.children.items():
            if val.isWord:
                words.append(val.full_word)
            res = self.all_words_from_node(val)
            words = words + res
        return words

    def give_possible_words(self, prefix):
        cur = self.root
        for letter in prefix:
            if letter not in cur.children:
                return ""

            cur = cur.children[letter]
        possibilities = self.all_words_from_node(cur)
        return possibilities


t = Trie()
with open("words/words.txt") as f:
    for word in f.readlines():
        t.add_word(word.strip())


while True:
    user_input = input("Word: ")
    # if user_input == "exit":
    #    break
    print(t.give_possible_words(user_input)[:5])
